### DISCUSSION TITLE
Support yarn's behavior for CLI lookups

### DISCUSSION THREAD

**Discussion by orta** [pnpm/pnpm#3191] [Ideas]
Hi folks, I was wondering about whether pnpm was interested in replicating the yarn's CLI experience for unknown commands.

Today if you install a new dependency which adds a binary, these binaries can be used with `pnpx` or `ppm exec -r`

```sh
pnpm add typescript -D

❯ pnpm tsc
 ERROR  Missing script: tsc
❯ pnpm exec tsc
 ERROR  The "pnpm exec" command currently only works with the "-r" option
❯ pnpm exec -r tsc

❯ pnpm tsc --init
 ERROR   ERROR  Unknown option: 'init'
```

vs Yarn's

```
yarn add typescript --dev
yarn tsc
yarn tsc --init
```

Yarn's missing CLI command props will check: 

```sh
 - built-in command names
 - matching scripts in package.json
 - binaries in ./node_modules/.bin/
```

Which makes writing instructions trivial because it's always one app to call and much easier to reason about because it's always just a prefix to the rest of your commands. E.g. `yarn eslint --fix`. 

Switching to this pattern isn't a breaking change (all the other commands can still work as-is) but it allows for people to migrate to the simpler CLI experience. Is there interest?

( also somewhat discussion for this in npm https://github.com/npm/rfcs/pull/279 )

**Comment by zkochan** [comment]
I think it is a good idea. Contributions are welcomed.

**Reply by amitgilad3** [reply]
Hi, is anybody tackling this issue or can i give it a try ??

**Reply by zkochan** [reply]
Go ahead

**Reply by zkochan** [reply]
I will probably look into it now. Seems like you did not work on it yet.

**Reply by zkochan** [reply]
Actually, I am not sure we should add this to the Node.js CLI. I added it to the [wrapper written in Rust](https://github.com/pnpm/pn), which runs much faster

**Comment by KSXGitHub** [comment]
Why didn't you just use `pnpx`?

**Comment by zkochan** [comment]
:ship: pnpm v6.6.0

**Reply by seahindeniz** [reply]
Congratulations on the update. I just test it with a script like `pnpm -r lint` and works as expected. ✨
May I ask, should it support `pnpm recursive [arbitary command]`  or is this something like that should not be supported by design?

**Reply by zkochan** [reply]
Probably it should work but not all variants work now. I work on a subsequent update
https://github.com/pnpm/pnpm/pull/3492
